pkg/network/protocols/http/shared_libraries.go

```
@@ -35,118 +37,213 @@ func (l *libPath) Bytes() []byte {
 }
 
-func newSOWatcher(procRoot string, perfHandler *ddebpf.PerfHandler, rules ...soRule) *soWatcher {
+func newSOWatcher(perfHandler *ddebpf.PerfHandler, rules ...soRule) *soWatcher {
 	allFilters := make([]string, len(rules))
 	for i, r := range rules {
 		allFilters[i] = r.re.String()
 	}
 
 	all := regexp.MustCompile(fmt.Sprintf("(%s)", strings.Join(allFilters, "|")))
 	return &soWatcher{
-		procRoot:   procRoot,
-		hostMount:  os.Getenv("HOST_ROOT"),
-		all:        all,
-		rules:      rules,
-		loadEvents: perfHandler,
+		procRoot:       util.GetProcRoot(),
+		all:            all,
+		rules:          rules,
+		loadEvents:     perfHandler,
+		processMonitor: monitor.GetProcessMonitor(),
 		registry: &soRegistry{
-			procRoot: procRoot,
-			byPath:   make(map[string]*soRegistration),
-			byInode:  make(map[uint64]*soRegistration),
+			byID:          make(map[pathIdentifier]*soRegistration),
+			byPID:         make(map[uint32]*soRegistration),
+			blocklistByID: make(map[pathIdentifier]struct{}),
 		},
 	}
 }
 
+type soRegistration struct {
+	pathID pathIdentifier
+	//mTime  uint64 TODO(nplanel) reload uprobes if the file changed (dnf upgrade)
+
+	refcount     int
+	unregisterCB func(pathIdentifier) error
+}
+
+func (r *soRegistration) Unregister() (cleanup bool) {
+	r.refcount--
+	if r.refcount > 0 {
+		return false
+	}
+	if r.unregisterCB != nil {
+		if err := r.unregisterCB(r.pathID); err != nil {
+			log.Debugf("unregisterCB %s : %s", r.pathID.String(), err)
+		}
+	}
+	return true
+}
+
+func newRegistration(pathID pathIdentifier, unregister func(pathIdentifier) error) *soRegistration {
+	return &soRegistration{
+		pathID:       pathID,
+		unregisterCB: unregister,
+		refcount:     1,
+	}
+}
+
+func (w *soWatcher) processExit(pid uint32) {
+	w.registry.Unregister(pid)
+}
+
 // Start consuming shared-library events
 func (w *soWatcher) Start() {
-	seen := make(map[seenKey]struct{})
-	w.sync()
+	thisPID, _ := util.GetRootNSPID()
```

why do you ignore the error?

---

there are no much to do there, I added log.Warn